dashboard/src/reducers/charts.ts

```diff
@@ -56,14 +59,26 @@ const chartsReducer = (
 ): IChartState => {
   switch (action.type) {
     case getType(actions.charts.requestCharts):
-      return { ...state, isFetching: true };
+      state.records.set(action.payload, false);
+      return { ...state, isFetching: true, records: state.records };
     case getType(actions.charts.requestChartsCategories):
       return { ...state, isFetching: true };
     case getType(actions.charts.receiveCharts):
+      // only update the records if the received page matches the requested one
+      const areMatchingPages = action.payload.page === state.page;
+      if (areMatchingPages) {
+        state.records.set(state.page, true); // mark the current page as successfully retrieved
+        state.records.set(state.page + 1, false); // mark the next page as pending (not undefined)
+      }
       return {
         ...state,
         isFetching: false,
-        items: action.payload,
+        hasFinished: action?.meta,
+        items: areMatchingPages
+          ? uniqBy([...state.items, ...action.payload.items], "id")
+          : state.items, // if pages don't match, ignore payload to avoid duplicates
+        page: action?.meta ? action.payload.page : action.payload.page + 1, // if action.meta==true, it's the last chunk
+        records: state.records,
```

Comment:
I think we can simplify all this if we store `items` as a map. If we store:

```
 items[action.payload.page] = action.payload.items
```

You will get unique items per page and, I think, you won't need the `records` since you already have a map.